File: src/Core/Model/Entity/FieldWithOrder.cs
namespace AppBoxCore;

/// <summary>
/// 带排序标志的实体成员标识,用于主键或索引等
/// </summary>
public struct FieldWithOrder : IBinSerializable
{
    public short MemberId { get; private set; }
    public bool OrderByDesc { get; private set; }

    public FieldWithOrder(short memberId, bool orderByDesc = false)
    {
        MemberId = memberId;
        OrderByDesc = orderByDesc;
    }

    public void WriteTo(IOutputStream ws)
    {
        ws.WriteShort(MemberId);
        ws.WriteBool(OrderByDesc);
    }

    public void ReadFrom(IInputStream rs)
    {
        MemberId = rs.ReadShort();
        OrderByDesc = rs.ReadBool();
    }
}

File: src/Dev.Web/src/AppBoxCore/Model/FieldWithOrder.ts
import {IBinSerializable, IInputStream, IOutputStream} from "@/AppBoxCore";

export class FieldWithOrder implements IBinSerializable {
    private _memberId: number;
    private _orderByDesc: boolean;

    public get MemberId(): number {
        return this._memberId;
    }

    public get OrderByDesc(): boolean {
        return this._orderByDesc;
    }

    public constructor();
    public constructor(memberId: number, orderByDesc?: boolean);
    public constructor(memberId?: number, orderByDesc?: boolean) {
        if (memberId != undefined) {
            this._memberId = memberId;
            this._orderByDesc = orderByDesc == undefined ? false : orderByDesc;
        }
    }

    public Clone(): FieldWithOrder {
        return new FieldWithOrder(this._memberId, this._orderByDesc);
    }

    WriteTo(bs: IOutputStream): void {
        bs.WriteShort(this.MemberId);
        bs.WriteBool(this.OrderByDesc)
    }

    ReadFrom(bs: IInputStream): void {
        this._memberId = bs.ReadShort();
        this._orderByDesc = bs.ReadBool();
    }
}